stellargraph/layer/node2vec.py

```diff
@@ -0,0 +1,195 @@
+        """
+        Sets node_num and multiplicity from the generator.
+        Args:
+             generator: The supplied generator.
+        """
+        if not isinstance(generator, (Node2VecNodeGenerator, Node2VecLinkGenerator)):
+            raise TypeError(
+                "Generator should be an instance of Node2VecNodeGenerator or Node2VecLinkGenerator"
+            )
+
+        self.multiplicity = generator.multiplicity
+        self.input_node_num = generator.graph.number_of_nodes()
+
+        if len(list(generator.graph.node_types)) > 1:
+            raise RuntimeError("Node2Vec called on graph with more than one node type.")
+
+    def _get_sizes_from_keywords(self, kwargs):
+        """
+        Sets node_num and input_feature_size from the keywords.
+        Args:
+             kwargs: The additional keyword arguments.
+        """
+        try:
+            self.input_node_num = kwargs["node_num"]
+            self.multiplicity = kwargs["multiplicity"]
+
+        except KeyError:
+            raise KeyError(
+                "Generator not provided; node_num and multiplicity must be specified."
+            )
+
+    def __call__(self, xin):
+        """
+        Construct node representations from node ids through a look-up table.
+
+        Args:
+            xin (Keras Tensor): Batch input node ids.
+
+        Returns:
+            Output tensor.
+        """
+        input_initializer = keras.initializers.RandomUniform(minval=-1.0, maxval=1.0)
+        input_embedding = Embedding(
+            self.input_node_num,
+            self.emb_size,
+            input_length=1,
+            name="input_embedding",
+            embeddings_initializer=input_initializer,
+        )
+        h_layer = input_embedding(xin)
+        h_layer = Reshape((self.emb_size,))(h_layer)
+        # K.squeeze(h_layer, axis=0)
+
+        return h_layer
+
+    def node_model(self):
+        """
+        Builds a Node2Vec model for node prediction.
+
+        Returns:
+            tuple: (x_inp, x_out) where ``x_inp`` is a Keras input tensor
+            for the Node2Vec model and ``x_out`` is the Keras tensor
+            for the Node2Vec model output.
+
+        """
+        # Create tensor inputs
+        x_inp = Input(shape=(1,))
+
+        # Output from Node2Vec model
+        x_out = self(x_inp)
+
+        return x_inp, x_out
+
+    def link_model(self):
+        """
+        Builds a Node2Vec model for link or node pair prediction.
+
+        Returns:
+            tuple: (x_inp, x_out) where ``x_inp`` is a list of Keras input tensors for (src, dst) nodes in the node pairs
+            and ``x_out`` is a list of output tensors for (src, dst) nodes in the node pairs.
+
+        """
+        # Expose input and output sockets of the model, for source node:
+        x_inp_src, x_out_src = self.node_model()
+
+        # Expose input and out sockets of the model, for target node:
+        x_inp_dst = Input(shape=(1,))
```

Comment:
Why is this different from line 146? Can we just do 
```
x_inp_dst, x_out_dst = self.node_model()
```
?

Reply:
We use different embedding tables. In the node_model, we use the input_embedding table which is used to generate node embedding vectors, here we use the output_embedding table   for context node look-up to predict its existence.

Reply:
Fixed